File: app/src/main/java/ru/e6atb/chat/BoundedScrollView.java
package ru.e6atb.chat;

import android.content.Context;
import android.widget.ScrollView;

final class BoundedScrollView extends ScrollView {
	private final int maxHeight;

	BoundedScrollView(Context context, int maxHeight) {
		super(context);
		this.maxHeight = maxHeight;
	}

	@Override
	protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
		int mode = MeasureSpec.getMode(heightMeasureSpec);
		int size = MeasureSpec.getSize(heightMeasureSpec);
		int height = mode == MeasureSpec.UNSPECIFIED ? maxHeight : Math.min(size, maxHeight);
		super.onMeasure(widthMeasureSpec, MeasureSpec.makeMeasureSpec(height, MeasureSpec.AT_MOST));
	}
}

File: app/src/main/kotlin/ru/e6atb/chat/BoundedScrollView.kt
package ru.e6atb.chat

import android.content.Context
import android.widget.ScrollView

class BoundedScrollView(context: Context, private val maxHeight: Int) : ScrollView(context) {
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val mode = MeasureSpec.getMode(heightMeasureSpec)
        val size = MeasureSpec.getSize(heightMeasureSpec)
        val height = if (mode == MeasureSpec.UNSPECIFIED) maxHeight else minOf(size, maxHeight)
        super.onMeasure(widthMeasureSpec, MeasureSpec.makeMeasureSpec(height, MeasureSpec.AT_MOST))
    }
}
